File: src/main/kotlin/leetCode/easy/MajorityElement.kt
package leetCode.easy

class MajorityElement {
    fun majorityElement(nums: IntArray): Int {
        var counter = 0
        var majorityElement = 0
        for (i in 0 until nums.size) {
            val number = nums[i]
            if (counter == 0) {
                majorityElement = number
                counter++
                continue
            }

            if (number == majorityElement) {
                counter++
            } else {
                counter--
            }
        }

        return majorityElement
    }
}

fun main(args: Array<String>) {
    println(MajorityElement().majorityElement(intArrayOf(3, 2, 3)))
    println(MajorityElement().majorityElement(intArrayOf(2, 2, 1, 1, 1, 2, 2)))
    println(MajorityElement().majorityElement(intArrayOf(2, 1, 2, 1, 2, 2)))
}

File: go/leetCode/easy/MajorityElement.go
package easy

func majorityElement(nums []int) int {
	n := len(nums)
	frequencies := make(map[int]int)

	result := 0
	for _, num := range nums {
		frequencies[num] = frequencies[num] + 1
		if frequencies[num] > n/2 {
			result = num
		}
	}

	return result
}
